rbs/TypeTranslator.cc

```diff
@@ -0,0 +1,418 @@
+            std::string pathNameStr(name->start, name->length);
+            auto pathNameConst = ctx.state.enterNameConstant(pathNameStr);
+            parent = ast::MK::UnresolvedConstant(loc, std::move(parent), pathNameConst);
+        }
+    }
+
+    rbs_constant_t *name = rbs_constant_pool_id_to_constant(fake_constant_pool, typeName->name->constant_id);
+    std::string_view nameStr(name->start, name->length);
+
+    if (typePath == nullptr || typePath->length == 0) {
+        if (nameStr == "Array" && isGeneric) {
+            return ast::MK::T_Array(loc);
+        } else if (nameStr == "Hash" && isGeneric) {
+            return ast::MK::T_Hash(loc);
+        } else if (nameStr == "Set" && isGeneric) {
+            return ast::MK::T_Set(loc);
+        } else if (nameStr == "Class" && isGeneric) {
+            return ast::MK::T_Class(loc);
+        } else if (hasTypeParam(ctx, typeParams, nameStr)) {
+            auto name = ctx.state.enterNameUTF8(nameStr);
+            return ast::MK::Send1(loc, ast::MK::T(loc), core::Names::typeParameter(), loc, ast::MK::Symbol(loc, name));
+        }
+    }
+
+    auto nameConstant = ctx.state.enterNameConstant(nameStr);
+
+    return ast::MK::UnresolvedConstant(loc, std::move(parent), nameConstant);
+}
+
+sorbet::ast::ExpressionPtr classInstanceType(core::MutableContext ctx,
+                                             std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                                             rbs_types_classinstance_t *node, core::LocOffsets loc) {
+    auto offsets = TypeTranslator::nodeLoc(loc, (rbs_node_t *)node);
+    auto argsValue = node->args;
+    auto isGeneric = argsValue != nullptr && argsValue->length > 0;
+    auto typeConstant = typeNameType(ctx, typeParams, node->name, isGeneric, offsets);
+
+    if (isGeneric) {
+        auto argsStore = Send::ARGS_store();
+        for (rbs_node_list_node *list_node = argsValue->head; list_node != nullptr; list_node = list_node->next) {
+            auto argType = TypeTranslator::toRBI(ctx, typeParams, list_node->node, loc);
+            argsStore.emplace_back(std::move(argType));
+        }
+
+        return ast::MK::Send(offsets, std::move(typeConstant), core::Names::squareBrackets(), offsets, argsStore.size(),
+                             std::move(argsStore));
+    }
+
+    return typeConstant;
+}
+
+sorbet::ast::ExpressionPtr classSingletonType(core::MutableContext ctx,
+                                              std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams,
+                                              rbs_types_classsingleton_t *node, core::LocOffsets loc) {
+    auto offsets = TypeTranslator::nodeLoc(loc, (rbs_node_t *)node);
+    auto innerType = typeNameType(ctx, typeParams, node->name, false, offsets);
+    return ast::MK::Send1(loc, ast::MK::T(loc), core::Names::classOf(), loc, std::move(innerType));
+}
```

Comment:
I take it that RBS has no syntax for `T.class_of(A)[...]`?

Reply:
Not yet, but there is a proposal: https://github.com/ruby/rbs/issues/1542